drivers/sensor/kx022/kx022.c

```diff
@@ -0,0 +1,601 @@
+			break;
+
+		default:
+			return -ENOTSUP;
+		}
+	} else {
+		switch ((enum sensor_channel_kx022)chan) {
+		case SENSOR_CHAN_KX022_MOTION:
+			kx022_motion_dir_get(val, data->sample_motion_dir);
+			break;
+		case SENSOR_CHAN_KX022_TILT:
+			kx022_tilt_pos_get(val, data->sample_tspp);
+			kx022_tilt_pos_get(val + 1, data->sample_tscp);
+			break;
+		default:
+			return -ENOTSUP;
+		}
+	}
+
+	return 0;
+}
+
+static int kx022_channel_get(const struct device *dev, enum sensor_channel chan,
+			     struct sensor_value *val)
+{
+	struct kx022_data *data = dev->data;
+
+	return kx022_get_channel(chan, val, data, data->gain);
+}
+
+static const struct sensor_driver_api kx022_api_funcs = {
+	.attr_set = kx022_attr_set,
+#ifdef CONFIG_KX022_TRIGGER
+	.trigger_set = kx022_trigger_set,
+#endif
+	.sample_fetch = kx022_sample_fetch,
+	.channel_get = kx022_channel_get,
+};
+
+static int kx022_init(const struct device *dev)
+{
+	const struct kx022_config *const cfg = dev->config;
+	struct kx022_data *data = dev->data;
+	uint8_t chip_id;
+	uint8_t val;
+
+	if (cfg->bus_init(dev) < 0) {
+		return -EINVAL;
+	}
+
+	if (data->hw_tf->read_reg(dev, KX022_REG_WHO_AM_I, &chip_id) < 0) {
+		LOG_DBG("Failed reading chip id");
+		return -EIO;
+	}
+
+	if (chip_id != KX022_VAL_WHO_AM_I) {
+		LOG_DBG("Invalid chip id 0x%x", chip_id);
+		return -EIO;
+	}
+
+	LOG_DBG("Chip id 0x%x", chip_id);
+
+	/* s/w reset the sensor */
+	if (data->hw_tf->update_reg(dev, KX022_REG_CNTL2, KX022_MASK_CNTL2_SRST,
+				    KX022_MASK_CNTL2_SRST) < 0) {
+		LOG_DBG("s/w reset fail");
+		return -EIO;
+	}
+
+	/* Arbitrary delay for the device to be ready after reset */
+	k_msleep(KX022_RESET_DELAY);
+
+	if (data->hw_tf->read_reg(dev, KX022_REG_WHO_AM_I, &chip_id) < 0) {
+		LOG_DBG("Failed reading chip id");
+		return -EIO;
+	}
+
+	if (chip_id != KX022_VAL_WHO_AM_I) {
+		LOG_DBG("Invalid chip id 0x%x", chip_id);
+		return -EIO;
+	}
+
+	LOG_DBG("Chip id 0x%x", chip_id);
+
+	/* Make sure the KX022 is stopped before we configure */
+	val = (cfg->resolution << KX022_CNTL1_RES_SHIFT) |
+	      (cfg->full_scale << KX022_CNTL1_GSEL_SHIFT);
+	if (data->hw_tf->write_reg(dev, KX022_REG_CNTL1, val) < 0) {
+		LOG_DBG("Failed CNTL1");
+		return -EIO;
+	}
+
+	/* Set KX022 default odr */
+	if (data->hw_tf->update_reg(dev, KX022_REG_ODCNTL, KX022_MASK_ODCNTL_OSA, cfg->odr) < 0) {
+		LOG_DBG("Failed setting odr");
+		return -EIO;
+	}
+
+#ifdef CONFIG_KX022_TRIGGER
+	if (kx022_trigger_init(dev) < 0) {
+		LOG_ERR("Failed to initialize triggers.");
+		return -EIO;
+	}
+#endif
+
+	/* Set Kx022 to Operating Mode */
+	kx022_mode(dev, KX022_OPERATING_MODE);
+
+	if (cfg->full_scale == KX022_FS_2G) {
+		data->gain = GAIN_XL;
+	} else if (cfg->full_scale == KX022_FS_4G) {
+		data->gain = 2 * GAIN_XL;
+	} else if (cfg->full_scale == KX022_FS_8G) {
+		data->gain = 4 * GAIN_XL;
+	}
+
+	/* After setting, need some delay, otherwise first polling data is wrong */
+	k_msleep(100);
```

Comment:
magic number: needs to be defined, justified. note that this will delay boot by 100ms.